crates/trie/sparse/src/state.rs

```diff
@@ -377,56 +384,50 @@ where
     ) -> SparseStateTrieResult<()> {
         let revealed_nodes = self.revealed_storage_paths.entry(account).or_default();
 
-        let FilteredProofNodes {
+        let FilterMappedProofNodes {
+            root_node,
             nodes,
             new_nodes,
             total_nodes: _total_nodes,
             skipped_nodes: _skipped_nodes,
-        } = filter_revealed_nodes(storage_subtree.subtree, revealed_nodes)?;
+        } = filter_map_revealed_nodes(
+            storage_subtree.subtree,
+            revealed_nodes,
+            &storage_subtree.branch_node_hash_masks,
+            &storage_subtree.branch_node_tree_masks,
+        )?;
         #[cfg(feature = "metrics")]
         {
             self.metrics.increment_total_storage_nodes(_total_nodes as u64);
             self.metrics.increment_skipped_storage_nodes(_skipped_nodes as u64);
         }
-        let mut nodes = nodes.into_iter().peekable();
 
-        if let Some(root_node) = Self::validate_root_node_decoded(&mut nodes)? {
+        if let Some(root_node) = root_node {
+            // Perform sanity check.
+            if matches!(root_node.node, TrieNode::EmptyRoot) && !nodes.is_empty() {
+                return Err(SparseStateTrieErrorKind::InvalidRootNode {
+                    path: root_node.path,
+                    node: alloy_rlp::encode(&root_node.node).into(),
+                }
+                .into())
+            }
+
             // Reveal root node if it wasn't already.
             let trie = self.storages.entry(account).or_default().reveal_root(
-                root_node,
-                TrieMasks {
-                    hash_mask: storage_subtree
-                        .branch_node_hash_masks
-                        .get(&Nibbles::default())
-                        .copied(),
-                    tree_mask: storage_subtree
-                        .branch_node_tree_masks
-                        .get(&Nibbles::default())
-                        .copied(),
-                },
+                root_node.node,
+                root_node.masks,
                 self.retain_updates,
             )?;
 
-            // Reserve the capacity for new nodes ahead of time.
+            // Reserve the capacity for new nodes ahead of time, if the trie implementation
+            // supports doing so.
             trie.reserve_nodes(new_nodes);
 
-            // Reveal the remaining proof nodes.
-            for (path, node) in nodes {
-                let (hash_mask, tree_mask) = if let TrieNode::Branch(_) = node {
-                    (
-                        storage_subtree.branch_node_hash_masks.get(&path).copied(),
-                        storage_subtree.branch_node_tree_masks.get(&path).copied(),
-                    )
-                } else {
-                    (None, None)
-                };
-
-                trace!(target: "trie::sparse", ?account, ?path, ?node, ?hash_mask, ?tree_mask, "Revealing storage node");
```

Comment:
re-adding a trace log showing num storage nodes being revealed would also be nice here